network/server.cpp

```diff
@@ -1,62 +1,77 @@
 #include "../include/network/server.hpp"
+#include "../include/network/DelayEstimator.hpp"
 
 /* PRIVATE */
 void NetworkServer::newConnection() {
   QTcpSocket *clientConnection = server->nextPendingConnection();
-  connect(clientConnection, &QAbstractSocket::disconnected, clientConnection, &QObject::deleteLater);
+  // The delete later is a problem for classes having a pointer to the client Socket (DelayEstimator)
+  // connect(clientConnection, &QAbstractSocket::disconnected, clientConnection, &QObject::deleteLater);
 
   std::string username = "";
   ServerConnection *client = new ServerConnection(username, clientConnection, this);
   clients.push_back(client);
   std::cout << "new client" << std::endl;
-
-  // clientConnection->write("Hello you !\n");
-  // clientConnection->disconnectFromHost();
 }
 
 void NetworkServer::sendJsonObject(std::string username, QJsonObject obj) {
   for (size_t i = 0; i < clients.size(); i++) {
     if (clients[i]->getUsername() == username) {
-      QJsonDocument doc = QJsonDocument(obj);
-      QByteArray msg = doc.toJson(JSON_FORMAT);
-      int size = msg.size();
-      std::cout << "SENDING (size: "<< msg.size() << " bytes)(to " << clients[i]->getUsername() << ")\n";
-      clients[i]->write(size);
-      clients[i]->write(msg);
+      clients[i]->write(obj);
       return;
     }
   }
   printf("Not such client found: %s\n", username.c_str());
 }
 
 void NetworkServer::broadcast(QJsonObject obj) {
-  QJsonDocument doc = QJsonDocument(obj);
-  QByteArray msg = doc.toJson();
-  int size = msg.size();
-  std::cout << "SENDING (size: "<< msg.size() << " bytes)\n";
   // for each element e of the Set clients
   for (size_t i = 0; i < clients.size(); i++) {
-    clients[i]->write(size);
-    clients[i]->write(msg);
+    clients[i]->write(obj);
   }
 }
 
+// Using the ping estimation so that clients receive the messages at the same time
+void NetworkServer::synchronizedBroadcast(QJsonObject obj) {
+  std::map<std::string, uint32_t> pings = delayEstimator->getPings();
+  uint32_t maxPing = delayEstimator->maxPing();
+  // std::vector<SendAfter*> senders;
+  for (size_t i = 0; i < clients.size(); i++) {
+    auto ping = pings.find(clients[i]->getUsername());
+    if (ping != pings.end()) {
+      SendAfter *s = new SendAfter(this, obj, maxPing - ping->second);
+      connect(s, &SendAfter::sendNow, clients[i], &ServerConnection::sendNow);
+      // senders.push_back();
+    } else {
+      std::cout << "Can't find "<< clients[i]->getUsername() << " in pings\n";
+    }
+  }
+  emit sendAfterStart(QThread::NormalPriority);
```

Comment:
We don't need this `QThread::Priority` param

Reply:
Yes artifact of QThread